### DISCUSSION TITLE
Allow sorting stacked time series from the legend

### DISCUSSION THREAD

**Discussion by amarant** [grafana/grafana#54142] [Dashboards, Panels & Visualizations]
So it seems the ability to sort stacked series via legend which was missing in the new Time Series panel was finally implemented in #35429 but what grafana developers giveth and then taketh away in #47373 and the only path forward given is the [Organize fields transform](https://grafana.com/docs/grafana/latest/panels/reference-transformation-functions/#organize-fields) which is a **manual** sorting (manual !!!).

**Sorting via legend is useful**, it is a way to explore the data, I understand that the main developer of the time series panel hates stacking, but stacking exist and won't go away, also stack ordering is a way to counteract all the possible misleading elements of stacking because it help you see the data in different ways. It is especially useful if you order by range so that series that varies less can be at the bottom.

Also it is not possible in SQL to send wide data frame with series ordered as we want with dynamic metrics, so it should be a grafana feature and by default grafana order lexicographically on the metric name (totally arbitrary). Also it was implemented, I just don't understand the back and forth here.

The reasons quoted here are :
> to avoid
> Series from re-stacking dynamically during data updates (can be quite bad experience with spikes, if stacking by dynamic min, max, last, or avg)
> Dashboards changing unpredictably from day to day or reload-to-reload

So maybe it shouldn't keep a legend order by default, but just a user exploratory step after the initial load. Or maybe just behind a checkbox named : "I understand it can be annoying but I really need the feature".

Please, please, please re-add the feature.

**Comment by leeoniya** [comment]
> I understand that the main developer of the time series panel hates stacking, but stacking exist and won't go away

heh, it's true, but i did not just make up the quoted reason as an excuse to remove legend-controlled stacking order. having the stacking order be dynamic based on legend sorting is a terrible experience, slow perfomance, and prevents internal refactoring we want to do in the coming months.

i agree with you that as an exploratory thing it is useful, but also very limited because, again, sorting by aggregate or instant stats doesnt provide a lot of control. ideally we would have the manual drag/drop sorting integrated into the legend and then that order can be "copied" and saved into an Organize fields transform that is then stable.

it is also useful for the stacking order to be different from legend order. you may want to sort legend dynamically by avg, but stack with least noisy series on the bottom. decoupling these concepts is fundamentally a good idea. unfortunately it does mean that this feature no longer works and we need to think of something else, but it's not straightforward, given that the legend can still be dynamic, but must also somehow provide static config adjustment.

**Reply by leeoniya** [reply]
> was finally implemented in https://github.com/grafana/grafana/issues/35429

looking at https://github.com/grafana/grafana/issues/35429 and the linked issues, it is absolutely not clear to me that stacking control was the primary motivation for that issue; it was almost certainly just being able to sort the legend itself. the fact that it affected stacking was likely done simply for parity, without giving much thought to the side-effects, which became apparent later.

> but what grafana developers giveth and then taketh away

i should add that we shipped this change 4 months ago in v8.5.0 to millions of users, and this is the first time someone noticed -- a strong indicator that this feature was more of an easter egg than something that absolutely had to remain in place.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
